### PR TITLE
[v2 rc] namespace refactor

### PR DESCRIPTION
**Pull request by xzilja**
## Description
Implements new `requiredNamespaces` and `namespaces` structure. Note, currently validation doesn't account for `extension` content, that will come in follow up PRs.

## Tests
```
  Client Validation
    connect
      ✓ throws when no params are passed
      ✓ throws when invalid pairingTopic is provided
      ✓ throws when empty pairingTopic is provided
      ✓ throws when non existant pairingTopic is provided
      ✓ throws when empty requiredNamespaces are provided
      ✓ throws when invalid requiredNamespaces are provided
      ✓ throws when no requiredNamespaces are provided
    pair
      ✓ throws when no params are passed
      ✓ throws when empty uri is provided
      ✓ throws when invalid uri is provided
      ✓ throws when no uri is provided
    approve
      ✓ throws when no params are passed
      ✓ throws when invalid id is provided
      ✓ throws when empty id is provided
      ✓ throws when no id is provided
      ✓ throws when invalid namespaces are provided
      ✓ throws when empty namespaces are provided
      ✓ throws when no namespaces are provided
      ✓ throws when invalid relayProtocol are provided
      ✓ throws when empty relayProtocol is provided
    reject
      ✓ throws when no params are passed
      ✓ throws when invalid id is provided
      ✓ throws when empty id is provided
      ✓ throws when no id is provided
      ✓ throws when empty reason is provided
      ✓ throws when invalid reason is provided
      ✓ throws when no reason is provided
      ✓ throws when invalid reason code is provided
      ✓ throws when empty reason code is provided
      ✓ throws when no reason code is provided
      ✓ throws when invalid reason message is provided
      ✓ throws when empty reason message is provided
      ✓ throws when no reason message is provided
    update
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when non existant topic is provided
      ✓ throws when invalid namespaces are provided
      ✓ throws when empty namespaces are provided
      ✓ throws when no namespaces are provided
    extend
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when non existant topic is provided
    request
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when non existant topic is provided
      ✓ throws when invalid chainId is provided
      ✓ throws when empty chainId is provided
      ✓ throws when chain id is not in session namespace
      ✓ throws when invalid request is provided
      ✓ throws when empty request is provided
      ✓ throws when no request is provided
      ✓ throws when invalid request method is provided
      ✓ throws when empty request method is provided
      ✓ throws when no request method is provided
      ✓ throws when request doesn't exist for given chainId
    respond
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when no response or error is passed
      ✓ throws when no id is passed
      ✓ throws when invalid id is passed
      ✓ throws when no jsonrpc is passed
      ✓ throws when invalid jsonrpc is passed
      ✓ throws when empty jsonrpc is passed
    ping
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when non existant topic is provided
    emit
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when non existant topic is provided
      ✓ throws when invalid chainId is provided
      ✓ throws when empty chainId is provided
      ✓ throws when invalid event is provided
      ✓ throws when empty event is provided
      ✓ throws when no event is provided
      ✓ throws when invalid event name is provided
      ✓ throws when empty event name is provided
      ✓ throws when no event name is provided
      ✓ throws when event doesn't exist for given chainId
    disconnect
      ✓ throws when no params are passed
      ✓ throws when invalid topic is provided
      ✓ throws when empty topic is provided
      ✓ throws when no topic is provided
      ✓ throws when non existant topic is provided


  Client Integration
    ✓ init
    connect
      ✓ connect (with new pairing) (44ms)
      ✓ connect (with old pairing)
    disconnect
      pairing
        ✓ deletes the pairing on disconnect
      session
        ✓ deletes the session on disconnect
    ping
      ✓ throws if the topic is not a known pairing or session topic
      pairing
        ✓ A pings B with existing pairing
        ✓ B pings A with existing pairing
        ✓ clients can ping each other after restart (60ms)
      session
        ✓ A pings B with existing session
        ✓ B pings A with existing session
        ✓ clients can ping each other after restart (65ms)
    update
      ✓ updates session namespaces state with provided namespaces
    extend
      ✓ updates session expiry state
```

### CODE REVIEW

**Comment by bkrem** [comment]
Linking this back to the related SDK Team PR: https://github.com/WalletConnect/sdk-team/pull/48

File: packages/client/src/constants/client.ts [hunk 1]
```diff
@@ -17,9 +17,8 @@ export const CLIENT_SHORT_TIMEOUT = 50;
 
 export const CLIENT_EVENTS: Record<ClientTypes.Event, ClientTypes.Event> = {
   session_proposal: "session_proposal",
-  update_accounts: "update_accounts",
-  update_namespaces: "update_namespaces",
-  update_expiry: "update_expiry",
+  update: "update",
```
Comment (pedrouid) [inline packages/client/src/constants/client.ts]:
```suggestion
  session_update: "session_update",
```

File: packages/client/src/constants/client.ts [hunk 2]
```diff
@@ -17,9 +17,8 @@ export const CLIENT_SHORT_TIMEOUT = 50;
...
+  extend: "extend",
```
Comment (pedrouid) [inline packages/client/src/constants/client.ts]:
```suggestion
  session_extend: "session_extend",
```

File: packages/client/test/client.spec.ts [hunk 3]
```diff
@@ -164,39 +164,22 @@ describe("Client Integration", () => {
...
         sessionA: { topic },
       } = await testConnectMethod(clients);
       const namespacesBefore = clients.A.session.get(topic).namespaces;
-      const namespacesAfter = [
+      const namespacesAfter = {
         ...namespacesBefore,
-        { chains: ["eip155:12"], methods: ["eth_sendTransaction"], events: ["accountsChanged"] },
-      ];
-      await clients.A.updateNamespaces({
+        eip000: {
```
Comment (pedrouid) [inline packages/client/test/client.spec.ts]:
```suggestion
        eip155: {
```

Comment (pedrouid) [reply packages/client/test/client.spec.ts]: Alternatively you can change the namespace is the accounts array to match

File: packages/types/src/client.ts [hunk 4]
```diff
@@ -11,9 +11,8 @@ import { IExpirer } from "./expirer";
 export declare namespace ClientTypes {
   type Event =
     | "session_proposal"
-    | "update_accounts"
-    | "update_namespaces"
-    | "update_expiry"
+    | "update"
```
Comment (pedrouid) [inline packages/types/src/client.ts]:
```suggestion
    | "session_update"
```

File: packages/types/src/client.ts [hunk 5]
```diff
@@ -11,9 +11,8 @@ import { IExpirer } from "./expirer";
...
+    | "extend"
```
Comment (pedrouid) [inline packages/types/src/client.ts]:
```suggestion
    | "session_extend"
```

File: packages/types/src/engine.ts [hunk 6]
```diff
@@ -16,9 +17,8 @@ export declare namespace EngineTypes {
   type Event =
     | "connect"
     | "approve"
-    | "update_accounts"
-    | "update_namespaces"
-    | "update_expiry"
+    | "update"
```
Comment (pedrouid) [inline packages/types/src/engine.ts]:
```suggestion
    | "session_update"
```

File: packages/types/src/engine.ts [hunk 7]
```diff
@@ -16,9 +17,8 @@ export declare namespace EngineTypes {
...
+    | "extend"
```
Comment (pedrouid) [inline packages/types/src/engine.ts]:
```suggestion
    | "session_extend"
```

File: packages/types/src/engine.ts [hunk 8]
```diff
@@ -16,9 +17,8 @@ export declare namespace EngineTypes {
   type Event =
     | "connect"
     | "approve"
```
Comment (pedrouid) [inline packages/types/src/engine.ts]:
```suggestion
    | "session_approve"
```

File: packages/types/src/engine.ts [hunk 9]
```diff
@@ -16,9 +17,8 @@ export declare namespace EngineTypes {
   type Event =
     | "connect"
```
Comment (pedrouid) [inline packages/types/src/engine.ts]:
```suggestion
    | "session_connect"
```

File: packages/types/src/session.ts [hunk 10]
```diff
@@ -3,24 +3,24 @@ import { RelayerTypes } from "./relayer";
 import { IStore } from "./store";
 
 export declare namespace SessionTypes {
-  type Accounts = string[];
-
   type Expiry = number;
 
-  interface Namespace {
+  interface NamespaceBody {
```
Comment (pedrouid) [inline packages/types/src/session.ts]:
```suggestion
  interface Namespace {
```

Comment (pedrouid) [reply packages/types/src/session.ts]: We could also have a BaseNamespace which is just methods and events and just extend RequiredNamespace and Namespace from it

That way it's easier to define an ExtensionNamespace without using Omit<_, _>

File: packages/types/src/session.ts [hunk 11]
```diff
@@ -32,8 +32,10 @@ export declare namespace SessionTypes {
   }
 
   interface Updatable {
-    accounts?: Accounts;
-    namespace?: Namespace;
+    namespace?: {
```
Comment (pedrouid) [inline packages/types/src/session.ts:38]: I mean I think we had decided on the SDK meeting that updates would require the whole namespaces

However if we want to make the update() API simpler wouldn't be easier to pass it as parameters?

update(topic, key, namespace)

Comment (pedrouid) [reply packages/types/src/session.ts:38]: On second thought, let's stick to the most basic API like it's defined on the `sdk-team`.

```typescript
interface Client {
    // for controller to update session namespaces
    public abstract update(params: {
      topic: string;
      namespaces: Map<string, SessionNamespace>;
    }): Promise<void>;
}
```

Comment (xzilja) [reply packages/types/src/session.ts:38]: This is not related to `update` method, this is used solely in `session.find`, I think better name for this interface is `Filters`

**Pull request merged**